samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatClient.java

```diff
@@ -56,7 +57,8 @@
 
   public ContainerHeartbeatClient(String coordinatorUrl, String executionEnvContainerId) {
     this.heartbeatEndpoint =
-        String.format("%scontainerHeartbeat?executionContainerId=%s", coordinatorUrl, executionEnvContainerId);
+        String.format("%s%s?%s=%s", coordinatorUrl, CoordinationConstants.CLUSTERBASED_CONTAINER_HEARTBEAT_SERVELET,
```

Comment:
I would suggest extracting the format as a constant and hardcoding some of the format constants instead of doing it the other way.
e.g. 
```
  CONTAINER_HEART_BEAT_SERVLET_FORMAT = "%s" + "/containerHeartbeat";
  CONTAINER_EXECUTION_ID_PARAM_FORMAT = "executionContainerId=" + "%s";
  CONTAINER_HEART_BEAT_ENDPOINT_FORMAT = CONTAINER_HEART_BEAT_SERVLET_FORMAT + CONTAINER_EXECUTION_ID_PARAM_FORMAT
```

Or some flavor of the above, so that you can reuse the format across the unit tests & other places in code. IMO, having the constants by themselves still forces users to figure out how they come together and repetitive code e.g. `"%s%s?%s=%s"` with readability hit as well.

Reply:
agree that "%s%s?%s=%s" is not readable. Will split as suggested above.
however, the point of the PR is to NOT have this hardcoded in both AM and container code separately. 
having them hardcoded separately in AM code and then have it exactly hardcoded in container code is risky imo